rpc/lib/server/ws_handler.go

```diff
@@ -0,0 +1,423 @@
+package rpcserver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"runtime/debug"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+
+	amino "github.com/tendermint/go-amino"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	types "github.com/tendermint/tendermint/rpc/lib/types"
+)
+
+///////////////////////////////////////////////////////////////////////////////
+// WebSocket handler
+///////////////////////////////////////////////////////////////////////////////
+
+// WebsocketManager provides a WS handler for incoming connections and passes a
+// map of functions along with any additional params to new connections.
+// NOTE: The websocket path is defined externally, e.g. in node/node.go
+type WebsocketManager struct {
+	websocket.Upgrader
+
+	funcMap       map[string]*RPCFunc
+	cdc           *amino.Codec
+	logger        log.Logger
+	wsConnOptions []func(*wsConnection)
+}
+
+// NewWebsocketManager returns a new WebsocketManager that passes a map of
+// functions, connection options and logger to new WS connections.
+func NewWebsocketManager(funcMap map[string]*RPCFunc, cdc *amino.Codec, wsConnOptions ...func(*wsConnection)) *WebsocketManager {
+	return &WebsocketManager{
+		funcMap: funcMap,
+		cdc:     cdc,
+		Upgrader: websocket.Upgrader{
+			CheckOrigin: func(r *http.Request) bool {
+				// TODO ???
+				return true
+			},
+		},
+		logger:        log.NewNopLogger(),
+		wsConnOptions: wsConnOptions,
+	}
+}
+
+// SetLogger sets the logger.
+func (wm *WebsocketManager) SetLogger(l log.Logger) {
+	wm.logger = l
+}
+
+// WebsocketHandler upgrades the request/response (via http.Hijack) and starts
+// the wsConnection.
+func (wm *WebsocketManager) WebsocketHandler(w http.ResponseWriter, r *http.Request) {
+	wsConn, err := wm.Upgrade(w, r, nil)
+	if err != nil {
+		// TODO - return http error
+		wm.logger.Error("Failed to upgrade to websocket connection", "err", err)
+		return
+	}
```

Comment:
Is it not perhaps better to do a `defer wsConn.Close()` after this `if` statement than to try to manage the closing of the WebSockets connection from within `wsConnection`?